baseplate/features/feature.py

```diff
@@ -0,0 +1,300 @@
+            "that is owned by <%s>. Please clean up.",
+            name,
+            feature_type,
+            owner,
+        )
+        return GloballyDisabledFeatureFlag()
+
+
+class FeatureFlagInterface(object):
+    """ Base interface for feature flag objects. """
+
+    def enabled(self, **args):
+        """ Return if the feature should be enabled for the given user and
+        targeting values.
+
+        :param baseplate.features.User user:
+        :param baseplate.featurs.TargetingParams targeting:
+        :rtype bool:
+        :return: True if the feature should be enabled, False if not.
+        """
+        raise NotImplementedError
+
+
+class GloballyEnabledFeatureFlag(FeatureFlagInterface):
+    """ A feature flag that is always enabled. """
+
+    def enabled(self, **args):
+        return True
+
+
+class GloballyDisabledFeatureFlag(FeatureFlagInterface):
+    """ A feature flag that is always disabled. """
+
+    def enabled(self, **args):
+        return False
+
+
+class FeatureTargeting(object):
+    """ Targeting configuration for a "basic" FeatureFlag. """
+
+    EXPECTED_USER_FLAGS = {"admin", "sponsor", "employee", "beta", "gold"}
+
+    def __init__(self, user_flags, newer_than, users, subreddits, subdomains,
+                 oauth_clients, url_flag):
+        assert not user_flags - self.EXPECTED_USER_FLAGS
+        self.user_flags = user_flags
+        self.newer_than = newer_than
+        self.users = users
+        self.subreddits = subreddits
+        self.subdomains = subdomains
+        self.oauth_clients = oauth_clients
+        self.url_flag = url_flag
+
+    @classmethod
+    def from_config(cls, config):
+        user_flags = config.get("user_flags", [])
+        user_flags = set([flag.lower() for flag in user_flags])
+        users = config.get("users", [])
+        users = set([user.lower() for user in users])
+        subreddits = config.get("subreddits", [])
+        subreddits = set([subreddit.lower() for subreddit in subreddits])
+        subdomains = config.get("subdomains", [])
+        subdomains = set([subdomain.lower() for subdomain in subdomains])
+        oauth_clients = config.get("oauth_clients", [])
+        oauth_clients = set([client.lower() for client in oauth_clients])
+        return cls(
+            user_flags=user_flags,
+            newer_than=config.get("newer_than"),
+            users=users,
+            subreddits=subreddits,
+            subdomains=subdomains,
+            oauth_clients=oauth_clients,
+            url_flag=config.get("url")
+        )
+
+
+class FeatureFlag(FeatureFlagInterface):
+    """ A "basic" feature flag.  Provides targeting against discrete values and
+    a percentage of users based on ID.
+    """
+
+    def __init__(self, id, name, owner, seed, percent_logged_in,
+                 percent_logged_out, targeting):
+        self.id = id
+        self.name = name
+        self.owner = owner
+        self.num_buckets = 1000
```

Comment:
My plan was to eventually have it configurable with the experiment config, but I wasn't sure if I wanted to open that up to start.